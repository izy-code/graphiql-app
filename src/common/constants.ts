export const MIN_PASSWORD_LENGTH = 8;
export const EMAIL_REGEX = /^([A-Z0-9_+-]+\.?)*[A-Z0-9_+-]@([A-Z0-9][A-Z0-9-]*\.)+[A-Z]{2,}$/i;
export const LOWERCASE_LETTER_REGEX = /\p{Ll}/u;
export const UPPERCASE_LETTER_REGEX = /\p{Lu}/u;
export const SPECIAL_CHARACTER_REGEX = /[-+:|/\\%*#@$!?^&]/;
export const NUMBER_REGEX = /\d/;

export const NO_ENDPOINT = 'no-endpoint-provided';
export const STORE_RESET = 'store-reset-action';
