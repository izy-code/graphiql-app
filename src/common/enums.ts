export enum LocalStorageKeys {
  URLS_RSS_REQUEST = 'REQUEST_LIST',
}

export enum NonProtectedPaths {
  SIGN_UP = '/sign-up',
  SIGN_IN = '/sign-in',
}

export enum ProtectedPaths {
  REST = '/GET',
  GRAPHIQL = '/graphiql',
  HISTORY = '/history',
}
