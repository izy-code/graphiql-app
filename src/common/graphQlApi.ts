'use server';

import { getIntrospectionQuery, type IntrospectionQuery } from 'graphql';

import type { TableRow } from '@/components/client-table/types.ts';

export interface GraphQLResponseData {
  status?: string;
  data?: object;
  errorMessage?: string;
}

export interface SchemaResponseData {
  data?: IntrospectionQuery;
  errorMessage?: string;
}

const createHeadersObject = (headers: TableRow[] = []): Record<string, string> =>
  headers.reduce(
    (acc, header) => {
      if (header.key.trim() && header.value.trim()) {
        acc[header.key] = header.value;
      }
      return acc;
    },
    {} as Record<string, string>,
  );

export const getSchema = async (endpoint: string, headers: TableRow[] = []): Promise<SchemaResponseData> => {
  try {
    if (!endpoint) {
      return { errorMessage: 'No schema endpoint provided' };
    }

    const response = await fetch(endpoint, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...createHeadersObject(headers),
      },
      body: JSON.stringify({ query: getIntrospectionQuery() }),
    });

    const responseBody = (await response.json()) as { data?: object; errors?: object };

    if ('errors' in responseBody && Array.isArray(responseBody.errors) && responseBody.errors.length > 0) {
      return { errorMessage: 'Response body contains errors' };
    }

    if (response.ok) {
      if ('data' in responseBody && responseBody.data && '__schema' in responseBody.data) {
        return { data: responseBody.data as IntrospectionQuery };
      }

      return {
        errorMessage: 'Unknown response structure: no "data" or "__schema" fields. Check endpoint and headers.',
      };
    }

    return { errorMessage: `Fetch failed with status code: ${response.status}` };
  } catch (error) {
    if (error instanceof Error) {
      if (error.message === 'fetch failed') {
        return { errorMessage: 'Failed to fetch schema' };
      }

      return { errorMessage: error.message };
    }

    return { errorMessage: 'Unknown error occurred while making the request' };
  }
};

export const getResponse = async (
  endpoint: string,
  query: string,
  variables = '{}',
  headers: TableRow[] = [],
): Promise<GraphQLResponseData> => {
  try {
    if (!endpoint || !query) {
      return { errorMessage: 'No endpoint or query provided' };
    }

    let parsedVariables = {};

    try {
      parsedVariables = JSON.parse(variables.trim() || '{}') as object;
    } catch (error) {
      return { errorMessage: 'Variables field is not valid JSON' };
    }

    const response = await fetch(endpoint, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...createHeadersObject(headers),
      },
      body: JSON.stringify({
        query,
        variables: parsedVariables,
      }),
    });

    const responseBody = (await response.json()) as { data?: object; errors?: object };

    if ('errors' in responseBody && Array.isArray(responseBody.errors) && responseBody.errors.length > 0) {
      return { errorMessage: 'Response body contains errors' };
    }

    if (response.ok) {
      if (!('data' in responseBody)) {
        return { errorMessage: 'Unknown response structure: no data field. Check endpoint and request params.' };
      }

      return { status: response.status.toString(), data: responseBody.data };
    }

    return { status: response.status.toString(), data: responseBody };
  } catch {
    return { errorMessage: 'Unknown error occurred while making the request', status: 'Fetch error' };
  }
};
