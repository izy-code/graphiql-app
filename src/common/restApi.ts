export interface RestResponseData {
  status?: string;
  data?: object;
  errorMessage?: string;
}

export async function getResponse(
  method: string,
  endpoint: string,
  headers: Record<string, string> = {},
  body?: string,
): Promise<RestResponseData> {
  try {
    if (!endpoint) {
      return { errorMessage: 'restApi.errors.endpoint' };
    }

    const options: RequestInit = {
      method,
      headers,
      cache: 'no-store',
    };

    if (method !== 'GET' && body) {
      options.body = JSON.stringify(JSON.parse(body));
    }
    const response = await fetch(endpoint, options);
    const responseBody = (await response.json()) as { data?: object; errors?: object };

    if ('errors' in responseBody && Array.isArray(responseBody.errors) && responseBody.errors.length > 0) {
      return { errorMessage: 'restApi.errors.body' };
    }

    return { status: response.status.toString(), data: responseBody };
  } catch {
    return { errorMessage: 'restApi.errors.unknown', status: 'restApi.errors.status' };
  }
}
