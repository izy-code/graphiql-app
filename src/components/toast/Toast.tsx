import { Flip, ToastContainer, type ToastTransition } from 'react-toastify';

export function Toast(): JSX.Element {
  return (
    <div className="Toast">
      <ToastContainer
        position="bottom-center"
        rtl={false}
        autoClose={3000}
        hideProgressBar={false}
        pauseOnFocusLoss
        draggable
        pauseOnHover
        transition={Flip as ToastTransition}
        newestOnTop={false}
        theme="dark"
        closeOnClick
        limit={3}
      />
    </div>
  );
}
