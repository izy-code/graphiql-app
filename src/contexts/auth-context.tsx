'use client';

import { type User } from 'firebase/auth';
import { useRouter } from 'next/navigation';
import type { ReactNode } from 'react';
import { createContext, useEffect, useState } from 'react';
import { toast } from 'react-toastify';

import { Loader } from '@/components/loader/Loader';
import { auth, logOut } from '@/firebase/firebase';
import { useScopedI18n } from '@/locales/client';

export interface UserImpl {
  stsTokenManager: {
    expirationTime: number;
  };
}

export type AuthContextType = User | null;

const TOKEN_CHECK_INTERVAL_MS = 60 * 1000;

export const AuthContext = createContext<AuthContextType>(null);

export function AuthProvider({ children }: { children: ReactNode }): ReactNode {
  const [user, setUser] = useState<User | null>(null);
  const [isLoading, setIsLoading] = useState(true);
  const router = useRouter();
  const translate = useScopedI18n('auth');

  useEffect(() => {
    const checkTokenExpiration = async (currentUser: User): Promise<void> => {
      const userImpl = currentUser.toJSON() as UserImpl;
      const { expirationTime } = userImpl.stsTokenManager;
      const currentTime = new Date().getTime();

      if (currentTime >= expirationTime) {
        await logOut();
        toast.info(translate('expired'));
        router.push('/');
      }
    };

    const unsubscribe = auth.onIdTokenChanged(async (changedUser) => {
      setUser(changedUser);
      setIsLoading(false);

      if (changedUser) {
        await checkTokenExpiration(changedUser);
      }
    });

    const intervalId = setInterval(async () => {
      if (auth.currentUser) {
        await checkTokenExpiration(auth.currentUser);
      }
    }, TOKEN_CHECK_INTERVAL_MS);

    return (): void => {
      unsubscribe();
      clearInterval(intervalId);
    };
  }, [router, translate]);

  return (
    <AuthContext.Provider value={user}>
      {isLoading ? <Loader loaderText={translate('loading')} /> : children}
    </AuthContext.Provider>
  );
}
