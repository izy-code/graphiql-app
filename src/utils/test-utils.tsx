import type { RenderOptions } from '@testing-library/react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { type PropsWithChildren, type ReactElement, useRef } from 'react';
import { Provider } from 'react-redux';

import type { AppStore, RootState } from '@/store/store';
import { setupStore } from '@/store/store';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
}

export function renderWithProvidersAndUser(
  ui: ReactElement,
  { preloadedState = {}, store = setupStore(preloadedState), ...renderOptions }: ExtendedRenderOptions = {},
): { store: AppStore; user: ReturnType<typeof userEvent.setup>; container: HTMLElement } {
  function Wrapper({ children }: PropsWithChildren): ReactElement {
    const storeRef = useRef<AppStore>();

    if (!storeRef.current) {
      storeRef.current = setupStore(preloadedState);
    }

    return <Provider store={storeRef.current}>{children}</Provider>;
  }

  return { store, user: userEvent.setup(), ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderWithUserSetup(ui: ReactElement): {
  user: ReturnType<typeof userEvent.setup>;
  container: HTMLElement;
} {
  return {
    user: userEvent.setup(),
    ...render(ui),
  };
}
