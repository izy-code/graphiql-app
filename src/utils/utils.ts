import {
  LOWERCASE_LETTER_REGEX,
  MIN_PASSWORD_LENGTH,
  NUMBER_REGEX,
  SPECIAL_CHARACTER_REGEX,
  UPPERCASE_LETTER_REGEX,
} from '@/common/constants';
import type { TableRow } from '@/components/client-table/types.ts';
import en from '@/locales/en';
import ru from '@/locales/ru';

const regexArray = [NUMBER_REGEX, UPPERCASE_LETTER_REGEX, LOWERCASE_LETTER_REGEX, SPECIAL_CHARACTER_REGEX];
export const MAX_STRENGTH = regexArray.length + 1;
export const getPasswordStrength = (password: string): number => {
  let strength = regexArray.reduce((result, regex) => result + Number(regex.test(password)), 0);
  strength += password.length >= MIN_PASSWORD_LENGTH ? 1 : 0;

  return strength;
};

export const translateText = (key: string): string =>
  window.location.pathname.split('/').includes('en') ? en[key as never] : ru[key as never];

export const encodeBase64 = (str: string): string => Buffer.from(str, 'utf-8').toString('base64').replace(/=+$/, '');

export const decodeBase64 = (str: string): string => {
  const pad = str.length % 4 === 0 ? '' : '===='.slice(str.length % 4);
  return Buffer.from(str + pad, 'base64').toString('utf-8');
};

export const generateUniqueId = (): string => `${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;

export const replaceVariables = (text: string | undefined, variables: TableRow[]): string => {
  const variableMap = Object.fromEntries(
    variables.filter(({ key, value }) => key.trim() && value.trim()).map(({ key, value }) => [key, value]),
  );

  let result = text || '';
  Object.entries(variableMap).forEach(([variableKey, variableValue]) => {
    const variablePlaceholder = `{{${variableKey}}}`;
    result = result.replace(new RegExp(variablePlaceholder, 'g'), variableValue);
  });

  return result.trim();
};

export const getEncodedHeaders = (headersParameter: TableRow[]): string => {
  if (!headersParameter || headersParameter.length === 0) {
    return '';
  }

  const encodedHeaders = headersParameter
    .filter(({ key, value }) => key.trim() && value.trim())
    .map(({ key, value }) => `${encodeURIComponent(key)}=${encodeURIComponent(value)}`);

  return `?${encodedHeaders.join('&')}`;
};

export const updateHistory = (
  locale: string,
  method: string,
  endpoint: string,
  body: string,
  headers: TableRow[],
  variables: TableRow[],
): void => {
  const encodedEndpoint = endpoint ? encodeBase64(endpoint) : encodeBase64(' ');
  const encodedBody = encodeBase64(replaceVariables(body, variables) || ' ');
  const encodedHeaders = getEncodedHeaders(headers);

  window.history.replaceState(null, '', `/${locale}/${method}/${encodedEndpoint}/${encodedBody}${encodedHeaders}`);
};
