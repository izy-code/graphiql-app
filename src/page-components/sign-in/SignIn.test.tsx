import { screen } from '@testing-library/react';
import mockRouter from 'next-router-mock';

import SignInPage from '@/app/[locale]/sign-in/page';
import { NonProtectedPaths } from '@/common/enums';
import { Toast } from '@/components/toast/Toast';
import { AuthProvider } from '@/contexts/auth-context';
import { renderWithProvidersAndUser } from '@/utils/test-utils';

const translate = vi.fn((arg: string) => {
  switch (arg) {
    case 'email':
      return 'Email';
    case 'password':
      return 'Password';
    case 'submit':
      return 'Sign in';
    default:
      return '';
  }
});

const translateValidation = vi.fn((arg: string) => {
  switch (arg) {
    case 'email.format':
      return 'Email must have valid format';
    case 'password.number':
      return 'Password must contain a number';
    default:
      return '';
  }
});

const translateAuth = vi.fn(() => 'Loading Firebase...');

vi.mock('@/locales/client', async (importOriginal) => {
  const actual = await importOriginal<object>();

  return {
    ...actual,
    useScopedI18n: vi.fn((arg: string) => {
      switch (arg) {
        case 'sign-in':
          return translate;
        case 'validation':
          return translateValidation;
        case 'auth':
          return translateAuth;
        default:
          return {};
      }
    }),
  };
});

describe('Sign in page', () => {
  const original = window.location;

  beforeAll(() => {
    Object.defineProperty(window, 'location', {
      value: { ...window.location, pathname: '/en' },
      writable: true,
    });
  });

  afterAll(() => {
    Object.defineProperty(window, 'location', { configurable: true, value: original });
  });

  it('renders correctly', () => {
    const { container } = renderWithProvidersAndUser(
      <AuthProvider>
        <SignInPage />
      </AuthProvider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('validation error messages should be visible and submit button must be disabled', async () => {
    const { user } = renderWithProvidersAndUser(
      <AuthProvider>
        <SignInPage />
      </AuthProvider>,
    );

    const email = screen.getByLabelText(/email/i);
    const password = screen.getByLabelText(/password/i);

    await user.type(email, 'email');
    await user.type(password, 'password');

    expect(screen.getByText('Email must have valid format')).toBeInTheDocument();
    expect(screen.getByText('Password must contain a number')).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /sign in/i })).toBeDisabled();
  });

  it('user will be redirected to main page after sign in', async () => {
    await mockRouter.push(NonProtectedPaths.SIGN_IN);

    const { user } = renderWithProvidersAndUser(
      <AuthProvider>
        <SignInPage />
        <Toast />
      </AuthProvider>,
    );

    const email = screen.getByLabelText(/email/i);
    const password = screen.getByLabelText(/password/i);
    const submit = screen.getByRole('button', { name: /sign in/i });

    await user.type(email, 'test@test.com');
    await user.type(password, 'testtestA1@');
    await user.click(submit);

    expect(mockRouter.asPath).toEqual('/');
    expect(screen.getByText(`Successfully signed in with test@test.com!`)).toBeInTheDocument();
  });
});
