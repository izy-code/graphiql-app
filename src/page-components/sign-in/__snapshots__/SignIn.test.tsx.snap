// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sign in page > renders correctly 1`] = `
<div>
  <div
    class="_page_476574"
  >
    <h1
      class="_title_476574"
    />
    <form
      class="_form_476574"
      name="react-hook-form"
      novalidate=""
    >
      <div
        class="_field_ba64c0"
      >
        <label
          class="_label_ba64c0"
          for=":r0:"
        >
          Email
        </label>
        <input
          aria-invalid="false"
          autocomplete="email"
          class="_input_ba64c0"
          id=":r0:"
          name="email"
          type="text"
        />
        <p
          class="_errorMessage_ba64c0"
        />
      </div>
      <div
        class="_field_ba64c0"
      >
        <label
          class="_label_ba64c0"
          for=":r1:"
        >
          Password
        </label>
        <input
          aria-invalid="false"
          autocomplete="current-password"
          class="_input_ba64c0"
          id=":r1:"
          name="password"
          type="password"
        />
        <p
          class="_errorMessage_ba64c0"
        />
      </div>
      <button
        class="_button_1c0f93 _button--primary_1c0f93 _submitButton_476574"
        disabled=""
        type="submit"
      >
        Sign in
      </button>
    </form>
  </div>
  ;
</div>
`;
